### ISSUE TITLE
React: Header component framework loses state (remounts???)

### ISSUE THREAD

**Issue by bdwade100** [ag-grid/ag-grid#4624]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of AG Grid'
[] support request => see 'Requesting Community Support'
```

**Providing a Reproducible Scenario**
Plunker: https://plnkr.co/plunk/sZhBz4k0krrXGnJI

First, thanks for an awesome grid library!

I have provided a minimal working example from our application in the plunker above. We use an "AgGridReactWrapper" to wrap all instances where we use ag-grid. The bug first appeared in version 25.2.0 and is still present in 25.3.0.

**Current behavior**
The plunker has a header component in the first column. This header component has a button that mimics a three-state button in our application. If I click this button, it rotates through the three states. Under the following conditions, the button state will get reset to the initial state of zero:
* Clicking the "Refresh" button (re-fetches the data)
* Clicking "Size Columns To Fit" / "Auto Size All Columns" button
* Clicking "Enable Floating Filter" / "Disable Floating Filter" button

**Expected behavior**
I expect the header component to retain its state for all the conditions above.

This partially worked in version 25.1.0. In that version, the following conditions worked as expected (the third condition did not work as expected):
* Clicking the "Refresh" button (re-fetches the data)
* Clicking "Size Columns To Fit" / "Auto Size All Columns" button

**Please tell us about your environment:**
* Operating system: Windows 10 Professional 64-bit (latest)
* IDE: Visual Studio Code
* HTTP server: Node with Express

**AG Grid version:**
* 26.0.0 (with reactUi=false; react header components are not currently allowed with reactUi=true)
* 25.3.0
* 25.2.0

**Browser:**
* "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36"
* "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:90.0) Gecko/20100101 Firefox/90.0"
* "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36 Edg/92.0.902.67"

**Language:**
* ES6/7

**Comment by makinggoodsoftware** [comment]
Hi,

Your ticket has been flagged as in_zendesk, it means that we recognize this as an issue worth investigating so we have moved it to our official support channel for customers (zendesk)

We will carry on there with the investigation and we will update you as soon as possible.

Thanks

**Comment by shuheb** [comment]
Hello @bdwade100!

Thank you for reporting this - much appreciated!

Based on your description we've now reproduced this issue as a bug.

We have added this bug to our backlog and we are tracking it with the following reference and description:

AG-5712 - [React] When updating columns via useState hook, the refresh method on Custom Header Component doesn't fire

We try to fix bugs from one release to the next, so this should be fixed in the next release or the one after if it was raised too close to the next release date.

See the next release date and whether this item is selected for implementation on the NEXT RELEASE tab of the product pipeline:

https://www.ag-grid.com/ag-grid-pipeline/

I've implemented a workaround you can use in the meantime: https://plnkr.co/edit/p9GQkE2HayWhtiM4 

Thanks again for bringing this up with us - note that I will be closing this issue, however feel free to reopen if the bug does persist after the fix.

Kind Regards,
Shuheb

**Comment by bdwade100** [comment]
Hi @shuheb --

Thanks for taking a look at my request and for providing a workaround so quickly! I have successfully implemented your workaround and everything appears to work as expected.

I will monitor that status of AG-5712 and will reopen this issue if I see the defect again after upgrading to a version after it's fixed.

Thanks again!
Brad

**Comment by kiril-matev** [comment]
Hello @bdwade100 

We've investigated this further with the product team and found that this is not a bug. Here's why.

Please see how React component refresh is described here: 
https://www.ag-grid.com/react-data-grid/component-cell-renderer/#component-refresh

In other words, if a refresh method is present the grid will call it. If it isn’t AND if immutableData is set to true then the grid will automatically refresh a renderer component (in a React way).

To sum it up, the hook component won’t automatically get refreshed unless:
A) a refresh method is supplied **or** 
B) immutableData is true

As explained above, this is expected behavior, we've closed this bug as rejected.

I hope this helps.

**Comment by bdwade100** [comment]
Hi @kiril-matev --

Thanks for your response!

As per your comment, I investigated how the React component refresh is described in the link you provided and the [additional information for how header component refresh works](https://www.ag-grid.com/react-data-grid/component-header/#header-component-interface-2). Like you mentioned, both indicate how to make the refresh work correctly and indicate that the issue I reported is not a bug. Thanks for clarifying the documentation.

One additional question if I may ask. I could be wrong; however, it seems from the documentation on [Registering Components](https://www.ag-grid.com/react-data-grid/components/) that I can register custom components:
1. By name.
2. Direct reference.

I can get my example to work when registering by name using the Plunker from @shuheb ; however, I cannot get it to work using a direct reference. Also, I cannot find any documentation that says I have to register by name, only that it's preferred.

Can I only use hooks if I register the component by name? Is this documented and I missed it?

Thanks!
Brad
